proposals/0057-events.md

```diff
@@ -0,0 +1,313 @@
+in this case. CPI for event logging would be only "cheaper" when using >4k
+event payloads but most events are <512b in length. This cost enforces
+theoretical limits to the amount of logs per transaction & block:
+Limit per 200k CU: 800kB
+Limit per 48M CU: 200MB
+
+### RPC-Client
+
+Clients should be able subscribe to events emitted with various filters:
+
+```
+rpc.onEvents(
+    callback: EventCallback,
+    commitment?: Commitment,
+    filters?: EventsFilter[]);
+```
+
+Each event emitted calls the callback, the context of the instruction that
+emitted the event is referenced through Transaction signature and indexes that
+uniquely identify the instruction. `getTransaction` can be used to fetch the
+remaining details like account keys and instruction data efficiently in case
+they are needed for indexing.
+
+```
+EventCallback: ((
+    message: Buffer,
+    instructionContext: {
+        signature: TransactionSignature,
+        outerInstructionIndex: number,
+        innerInstructionIndex?: number,
+    }
+    context: {
+        slot: number
+    }
+) => void)
+```
+
+DataSizeFilter and MemcmpFilter can be used to filter for specific event types
+using their discriminator. AccountMetaFilter allows to restrict on events
+emitted from transactions reading or writing from/to specific Accounts.
+Similarly it allows to restricted based on specific signers as well on
+executing certain programs.
+SignatureFilter allows to restrict to events emitted by an individual
+transaction, while waiting for it's confirmation.
+
+```
+EventsFilter: (DataSizeFilter|MemcmpFilter|AccountMetaFilter|SignatureFilter)
+DataSizeFilter: {
+    dataSize: number;
+}
+
+MemcmpFilter: {
+    memcmp: {
+        bytes: string;
+        offset: number;
+    };
+}
+AccountMetaFilter : {
+    pubkey: Pubkey,
+    writeable?: bool,
+    signer?: bool,
+    program? bool,
+}
+SignatureFilter : {
+    signature: TransactionSignature
+}
+```
+
+Clients should be able to query signatures that emitted events with various
+filters. Any query must be constrained to a single programId to allow for
+efficient indexing. Options must allow to paginate the result so that response
+size and latency for programs with a lot of confirmed transactions can be
+controlled.
+
+```
+EventLookupFilter: (DataSizeFilter|MemcmpFilter|AccountMetaFilter)
+
+rpc.getSignaturesForEvent(
+    programId: PublicKey,
+    filters?: EventLookupFilter[],
+    options?: {
+        commitment?: Commitment,
+        minContextSlot?: number,
+        limit?: number,
+        before?: TransactionSignature,
+        until?: TransactionSignature
+    }
+): Promise<TransactionSignature[]>
+```
+
+In addition clients should be able to query events directly and index them
+reliably working around the randomized execution of entries inside a slot.
+The response is sorted by the following priorities:
+
+1. slots ascending
+2. signature order for transactions inside a block
```

Comment:
Rather than trying to order by signature here, there is a single natural ordering of how transactions are placed in entries and then in a block. This order is common across the cluster (`solana-ledger-tool` outputs it). During replay, the entry transactions are intentionally shuffled randomly but if that shuffling didn't exist then the entire cluster would replay entries the same way

Reply:
Oh wasn't aware, if we can keep that alive across replay, we should be able to have even more easy indexing. Will adapt.